### ISSUE TITLE
Missing manual?

### ISSUE THREAD

**Issue by 1player** [flathub/com.kristianduske.TrenchBroom#6]
Trying to open the Trenchbloom manual from the menu bar doesn't do anything at all. Is it correctly included in this package?

**Comment by Sinihopea** [comment]
> Trying to open the Trenchbloom manual from the menu bar doesn't do anything at all. Is it correctly included in this package?

Sorry but I don't think it is included. If I recall correctly I didn't include the documentation because it blocks the build process of Trenchbroom. Furthermore it caused other smaller issues which I was unable to solve. To build the documentation and include it into Flatpak, you have to bootstrap Glasgow Haskell Compiler (GHC) with all of its dependencies and then build Pandoc and its dependencies inside Flatpak build process. I found that task to be incredibly challenging.

People who use Pandoc with their Flatpaks typically use pre-compiled Pandoc binary or pre-compiled Haskell compiler they manually inject inside Flatpak but that will then limit the available build architectures to just one. I opted not to do that so that at some point when they release 2022.1 or so I could try and make the Trenchbroom also available for people using other than x86_64 architecture. Mainly 32-bit Intel architecture but possibly also aarch64 if it can be done.

The documentation itself consists of HTML documents forming a single-page doc. What clicking that manual item in menu would do is to open the documentation into your web browser. There was some discussion about this and it might cause a security issue since to make it work might require opening yet another portal into the Flatpak.

And finally Trenchbroom manual receives frequent updates faster than Trenchbroom receives stable numbered releases. So what all of this boils down to is that it takes massive amounts of work to include what is outdated documentation. People should read the documentation from here...

https://trenchbroom.github.io/manual/latest/

...and at the core the experience is identical except that now you have more up to date documentation. It's a completely different matter if the documentation has to be available for offline use. But since it can be easily downloaded with simple push of a Ctrl+S in most popular web browsers I personally don't think it's a big issue either.

I am still aware that Kristian wished for me to include the docs and I'd love to. But I don't think it's economical or ecological (bootstrapping Haskell compiler even on your local machine is a lot of work and it takes a long time to build).

EDIT: If I'm talking BS or missing something here I'd love to hear about it. I'm open to learning how to do this the proper way. I'll keep this thread open for now if someone with more experience than I have with Flatpak build process and Pandoc can assist with this issue.

**Comment by Sinihopea** [comment]
Well, I checked the Flatpak build receipt and we are including x86_64 version of pre-compiled Pandoc with the Flatpak build process. It might be that there is some other issue in play since that should mean that the documentation is compiled and should probably be available within the Flatpak. Perhaps the Flatpak wants to open a Firefox but is unable to do so and that blocks the manual from opening.

I'll try to look into it again come next weekend but I can't promise that I do.

**Comment by Calinou** [comment]
In the meantime, could the Flatpak have a patch to modify TrenchBroom's Open manual button so that it opens the default web browser pointing to `https://trenchbroom.github.io/manual/latest/`?

Regarding permissions, the Godot Flatpak is able to open the user's default web browser using the following permissions. Godot doesn't use Qt to achieve this though, it uses its own platform code:

https://github.com/flathub/org.godotengine.Godot/blob/335f5e69f4dcdb9a1cb7fc2ca8a9832a7ae825b2/org.godotengine.Godot.yaml#L42-L49

**Comment by Sinihopea** [comment]
> In the meantime, could the Flatpak have a patch to modify TrenchBroom's Open manual button so that it opens the default web browser pointing to `https://trenchbroom.github.io/manual/latest/`?

Yes, and this would be ideal. But it's up to Trenchbroom's developers to implement that.

> Regarding permissions, the Godot Flatpak is able to open the user's default web browser using the following permissions. Godot doesn't use Qt to achieve this though, it uses its own platform code:
> 
> https://github.com/flathub/org.godotengine.Godot/blob/335f5e69f4dcdb9a1cb7fc2ca8a9832a7ae825b2/org.godotengine.Godot.yaml#L42-L49

IIRC it's possible to implement call to "org.freedesktop.portal.OpenURI.OpenURI" which would then open user's web-browser and it's not tied to Qt. But it's up to Trenchbroom's developers to implement it. If it's implemented, I can then grant the permission for it to work from within the Flatpak.

**Comment by kduske** [comment]
> But it's up to Trenchbroom's developers to implement it. If it's implemented, I can then grant the permission for it to work from within the Flatpak.

Just for my understanding: It's up to me because you don't know how? Because I would think it would have to be patched in this repo rather than upstream.

**Comment by Sinihopea** [comment]
> > But it's up to Trenchbroom's developers to implement it. If it's implemented, I can then grant the permission for it to work from within the Flatpak.
> 
> Just for my understanding: It's up to me because you don't know how? Because I would think it would have to be patched in this repo rather than upstream.

I know how to patch it. But I'm not sure what's the Flathub's policy on making change this deep and how to take time to then maintain that patch. Usually these repositories just patch the build system if needed. Also it's not entirely trivial what approach should be taken.

The reason I wrote that it's ideal to make this change in upstream is that the manual that comes with the release is most of the time outdated in similar projects. And since people need external web browser in any case to view it they can just the same go take look at the most up-to-date manual at the web page you host it. It would be more beneficial to your users in my humble opinion. Shipping offline copy of the manual can stay there if that is wished as a backup when there is no internet connectivity.

However Flatpak's sandbox does not allow viewing of the documentation inside the package easily as far as I know. Options as far as I understand and I could be wrong... (Someone who is more familiar with things and up-to-date with latest features should clarify.)

- Building a HTML viewer into Trenchbroom for the manual using QWebView component (difficult since it doesn't come with Qt any longer and it's replacement QtWebEngineView IIRC doesn't support Qt Widgets and only works with Qt Quick). The amount of work and added dependencies makes this a bad idea in my opinion. But would allow viewing of the offline documentation within Flatpak. However it also might introduce security issues and break Flatpak's security model. It could be separate binary inside the package probably.
- Patch the QAction method you have in menus to make a DBus call to open external web browser through xdg-portal to then open the following URL: https://trenchbroom.github.io/manual/latest/ and users need to have xdg-desktop-portal installed as well in their host system from https://github.com/flatpak/xdg-desktop-portal

The one thing that can't be done / shouldn't be done according to best of my knowledge is:

- Make HTML viewer that is located outside Flatpak package to see into inside the package and open HTML file from there.

Regardless, no average user is able to easily view the documentation that is packaged within the Trenchbroom Flatpak right now. Like I explained year or two ago, it just as well could be left outside to make Trenchbroom's Flatpak smaller to distribute and easier to build and package. And have aarch64 (ARM) and 32-bit x86 based builds of Trenchbroom Flatpak available as well. But I've left it there since you asked me to do so.

**Comment by kduske** [comment]
At this time, I don't see a compelling reason to stop shipping the manual the way we do and then to just point a web browser to its location. It works on all platforms except for this one, so far. As long as that doesn't change, I'd like to keep it the way it is.

In my opinion, it is the responsibility of the maintainers of downstream port projects to patch the code to make it compatible with their platform specific requirements. Such patches should only be done upstream if they are beneficial for other platforms (or for me) or if there is no way of fixing the problem downstream.

> The reason I wrote that it's ideal to make this change in upstream is that the manual that comes with the release is most of the time outdated in similar projects.

That is not true for TrenchBroom. The online manual is always identical to the one in the latest release version. Conversely though, using the online manual condemns you to keeping the flatpak port up to date because you will be pointing users of older versions to the manual of the latest version, since we currently don't have version specific URLs for the online manual.

I'm not sure what the correct solution for flatpak would be given these restrictions. I could investigate keeping older versions of the manual online if you want to point users of the flatpak version to the online manual. I think keeping the older versions of the manual available makes sense in its own right.
